src/librustc/infer/error_reporting/anon_anon_conflict.rs

```diff
@@ -0,0 +1,197 @@
+use hir;
+use infer::InferCtxt;
+use ty::{self, Region};
+use infer::region_inference::RegionResolutionError::*;
+use infer::region_inference::RegionResolutionError;
+use hir::map as hir_map;
+use middle::resolve_lifetime as rl;
+use hir::intravisit::{self, Visitor, NestedVisitorMap};
+
+// The visitor captures the corresponding `hir::Ty` of the
+// anonymous region.
+struct FindNestedTypeVisitor<'a, 'gcx: 'a + 'tcx, 'tcx: 'a> {
+    infcx: &'a InferCtxt<'a, 'gcx, 'tcx>,
+    hir_map: &'a hir::map::Map<'gcx>,
+    bound_region: ty::BoundRegion,
+    found_type: Option<&'gcx hir::Ty>,
+}
+
+impl<'a, 'gcx, 'tcx> Visitor<'gcx> for FindNestedTypeVisitor<'a, 'gcx, 'tcx> {
+    fn nested_visit_map<'this>(&'this mut self) -> NestedVisitorMap<'this, 'gcx> {
+        NestedVisitorMap::OnlyBodies(&self.hir_map)
+    }
+
+    fn visit_ty(&mut self, arg: &'gcx hir::Ty) {
+        // Find the index of the anonymous region that was part of the
+        // error. We will then search the function parameters for a bound
+        // region at the right depth with the same index.
+        let br_index = match self.bound_region {
+            ty::BrAnon(index) => index,
+            _ => return,
+        };
+
+        match arg.node {
+            hir::TyRptr(ref lifetime, _) => {
+                match self.infcx.tcx.named_region_map.defs.get(&lifetime.id) {
+                    // the lifetime of the TyRptr!
+                    Some(&rl::Region::LateBoundAnon(debruijn_index, anon_index)) => {
+                        if debruijn_index.depth == 1 && anon_index == br_index {
+                            self.found_type = Some(arg);
+                            return; // we can stop visiting now
+                        }
+                    }
+                    Some(&rl::Region::Static) |
+                    Some(&rl::Region::EarlyBound(_, _)) |
+                    Some(&rl::Region::LateBound(_, _)) |
+                    Some(&rl::Region::Free(_, _)) |
+                    None => {
+                        debug!("no arg found");
+                    }
+                }
+            }
+            _ => {}
+        }
+        // walk the embedded contents: e.g., if we are visiting `Vec<&Foo>`,
+        // go on to visit `&Foo`
+        intravisit::walk_ty(self, arg);
+    }
+}
+
+impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
+    // This function calls the `visit_ty` method for the parameters
+    // corresponding to the anonymous regions. The `nested_visitor.found_type`
+    // contains the anonymous type.
+    pub fn find_anon_type(&self, region: Region<'tcx>, br: &ty::BoundRegion) -> Option<&hir::Ty> {
+        if self.is_suitable_anonymous_region(region).is_some() {
+            let def_id = self.is_suitable_anonymous_region(region).unwrap();
+            let node_id = self.tcx.hir.as_local_node_id(def_id).unwrap();
+            let ret_ty = self.tcx.type_of(def_id);
+            if let (&ty::TyFnDef(_, _), hir_map::NodeItem(it)) =
```

Comment:
Nit: I find the deep nesting a bit hard to read. You might "un-nested" by using early return; something like:

```rust
let it = match (&ret_ty.sty, self.tcx.hir.get(node_id)) {
  (&ty::TyFnDef(_, _), hir_map::NodeItem(it)) => it,
  _ => return None,
};
let fndecl = match it.node {
  hir::ItemFn(ref fndecl, ..) => fndecl,
  _ => return None,
};
fndecl
                        .inputs
                        .iter()
                        .filter_map(|arg| {
                            let mut nested_visitor = FindNestedTypeVisitor {
                                infcx: &self,
                                hir_map: &self.tcx.hir,
                                bound_region: *br,
                                found_type: None,
                            };
                            nested_visitor.visit_ty(&**arg);
                           if nested_visitor.found_type.is_some() {
                                nested_visitor.found_type
                            } else {
                                None
                            }
                        })
                        .next()
```